### PR TITLE
More informative error message for DataCollatorForSeq2Seq

### LINKED ISSUE #15505: ValueError: Unable to create tensor, you should probably activate truncation and/or padding with 'padding=True' 'truncation=True' to have batched tensors with the same length.
Maybe  @SaulLu can help?

## Information

I am following the [text summarization](https://huggingface.co/course/chapter7/5) tutorial on hugging face website which uses the mt5-small model. It explains step by step on how to perform a text summarization task. 

## To reproduce

1. Run the [following notebook](https://colab.research.google.com/github/huggingface/notebooks/blob/master/course/chapter7/section5_pt.ipynb)
2. cell # 32 should reproduce the following error. (it did for me)

```
ValueError                                Traceback (most recent call last)
File ~/PycharmProjects/nlp-env/lib/python3.9/site-packages/transformers/tokenization_utils_base.py:707, in BatchEncoding.convert_to_tensors(self, tensor_type, prepend_batch_axis)
    706 if not is_tensor(value):
--> 707     tensor = as_tensor(value)
    709     # Removing this for now in favor of controlling the shape with `prepend_batch_axis`
    710     # # at-least2d
    711     # if tensor.ndim > 2:
    712     #     tensor = tensor.squeeze(0)
    713     # elif tensor.ndim < 2:
    714     #     tensor = tensor[None, :]

ValueError: too many dimensions 'str'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
Input In [72], in <module>
----> 1 data_collator(features)

File ~/PycharmProjects/nlp-env/lib/python3.9/site-packages/transformers/data/data_collator.py:586, in DataCollatorForSeq2Seq.__call__(self, features, return_tensors)
    583         else:
    584             feature["labels"] = np.concatenate([remainder, feature["labels"]]).astype(np.int64)
--> 586 features = self.tokenizer.pad(
    587     features,
    588     padding=self.padding,
    589     max_length=self.max_length,
    590     pad_to_multiple_of=self.pad_to_multiple_of,
    591     return_tensors=return_tensors,
    592 )
    594 # prepare decoder_input_ids
    595 if (
    596     labels is not None
    597     and self.model is not None
    598     and hasattr(self.model, "prepare_decoder_input_ids_from_labels")
    599 ):

File ~/PycharmProjects/nlp-env/lib/python3.9/site-packages/transformers/tokenization_utils_base.py:2842, in PreTrainedTokenizerBase.pad(self, encoded_inputs, padding, max_length, pad_to_multiple_of, return_attention_mask, return_tensors, verbose)
   2839             batch_outputs[key] = []
   2840         batch_outputs[key].append(value)
-> 2842 return BatchEncoding(batch_outputs, tensor_type=return_tensors)

File ~/PycharmProjects/nlp-env/lib/python3.9/site-packages/transformers/tokenization_utils_base.py:212, in BatchEncoding.__init__(self, data, encoding, tensor_type, prepend_batch_axis, n_sequences)
    208     n_sequences = encoding[0].n_sequences
    210 self._n_sequences = n_sequences
--> 212 self.convert_to_tensors(tensor_type=tensor_type, prepend_batch_axis=prepend_batch_axis)

File ~/PycharmProjects/nlp-env/lib/python3.9/site-packages/transformers/tokenization_utils_base.py:723, in BatchEncoding.convert_to_tensors(self, tensor_type, prepend_batch_axis)
    718         if key == "overflowing_tokens":
    719             raise ValueError(
    720                 "Unable to create tensor returning overflowing tokens of different lengths. "
    721                 "Please see if a fast version of this tokenizer is available to have this feature available."
    722             )
--> 723         raise ValueError(
    724             "Unable to create tensor, you should probably activate truncation and/or padding "
    725             "with 'padding=True' 'truncation=True' to have batched tensors with the same length."
    726         )
    728 return self

ValueError: Unable to create tensor, you should probably activate truncation and/or padding with 'padding=True' 'truncation=True' to have batched tensors with the same length.

```

### PR DESCRIPTION
**Pull request by CakeCrusher**
# What does this PR do?
I ran into an error related to an incorrect shape of inputs when using `DataCollatorForSeq2Seq`. I learned that it had to do with the `BatchEncoding` class. I did not find the error message particularly helpful as  it does not mention anything about the input shape. Therefore I added the extra line on the error message to help guide anyone else who runs into this error.

Fixes #15505
@stas00 

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by CakeCrusher** [comment]
@stas00 agreed ill make a check for it

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -715,15 +715,14 @@ def convert_to_tensors(
 
                     self[key] = tensor
             except:  # noqa E722
-                if key == "overflowing_tokens":
+                if key == "overflowing_tokens" or key == "input_ids" or key == "attention_mask":
```
Comment (CakeCrusher) [inline src/transformers/tokenization_utils_base.py:718]: I'm not sure if there are more keys to take take into account for this issue

File: src/transformers/tokenization_utils_base.py [hunk 2]
```diff
@@ -715,15 +715,14 @@ def convert_to_tensors(
...
                     raise ValueError(
-                        "Unable to create tensor returning overflowing tokens of different lengths. "
+                        f"Unable to create tensor returning {key} of different lengths. "
                         "Please see if a fast version of this tokenizer is available to have this feature available."
```
Comment (CakeCrusher) [inline src/transformers/tokenization_utils_base.py:721]: Not sure if this feature of fast tokenizers applies for the other keys.

File: src/transformers/tokenization_utils_base.py (hunk 1 above)
Comment (stas00) [reply src/transformers/tokenization_utils_base.py:718]: I think there was a miscommunication there.

In the OP you suggested that there could be an error of passing wrongly shaped inputs - at least that's how I understood it. And I proposed that perhaps it'd be better to check if inputs are misformatted and assert if that's the case. But your newly proposed change is something totally different. So I think I lost you here.

Perhaps you could show an example of wrongly shaped inputs and then the corrected one?

**Comment by CakeCrusher** [comment]
Hi @stas00 , is the pr ok?

**Comment by stas00** [comment]
This PR is waiting for your answer here: https://github.com/huggingface/transformers/pull/17447#discussion_r883964302

**Comment by stas00** [comment]
@CakeCrusher, I think we lost each other here. Should we finish this PR?

**Comment by CakeCrusher** [comment]
@sgugger @stas00

Hi @stas00 sorry for the discontinuity, I am now able to focus and see this issue through.

Here is an example demonstrating successful and erring inputs:
https://colab.research.google.com/drive/16aLu6QrDSV_aUYRdpufl5E4iS08qkUGj?usp=sharing

I then made the following changes to overcome excessive nesting (a list containing a single item):
https://github.com/CakeCrusher/transformers/compare/main...lead_nesting_solution

I understand the changes are pretty fundamental, but they work. I have yet to add the assert statement, since the nesting fix does the job forcefully. I was hoping to do an overarching PR, involving the new error message (or assert) and the fix (possibly parametrized so that it is not forced). What are your thoughts?

**Comment by stas00** [comment]
This is an interesting idea, but I'm concerned it might be (1) not backward compatible (2) I think it's best for the user to apply this function themselves. Perhaps if it's useful util function we can provide it and assert with a message to use it instead?

And to remind my initial suggestion was:

- check if shape is wrong and raise a specific assert if it is wrong (with possible hints at how to fix it)

e.g. the inputs shape is wrong, expecting a, but got b....

won't that be a clean solution?

we can then discuss with others if they feel your proposed util function would be a good match to add.

**Comment by CakeCrusher** [comment]
@stas00 

> Perhaps if it's a useful util function we can provide it and assert with a message to use it instead?

That is an excellent idea.

I will have it ready early next week with a test.

Do you recommend I make a new PR for it or merge it to this one?

**Comment by CakeCrusher** [comment]
Hi @stas00 , I submitted a [new PR](https://github.com/huggingface/transformers/pull/18116) for the aforementioned fixes. I have yet to add the test and proper docs. As for what I have so far please let me know what you think.


(My git tree was a mess, so that was largely why it's a new PR sorry about that.)

**Comment by stas00** [comment]
Apologies for taking a long time to follow up, @CakeCrusher 

As I suggested in the first place I think your suggestion to assert on invalid input nesting is great.

I see you tried to move the helper util to `datasets` and it's not being welcomed there, as it's really a user's responsibility to prepare the data correctly.

Perhaps we just stick to the assert part and trust the user to figure out how to fix it?

@sgugger, are you ok with the assertion on the invalid input? I'd guess that you too might be against the 2nd part of adding a helper util to remove excessive nesting.

**Comment by CakeCrusher** [comment]
No worries @stas00,
Yeah.. I understand if I have to give up on introducing the helper function on this PR. I'll see what what [lhoestq](https://github.com/lhoestq) ends up thinking about the datasets alternative.

In the meantime, I'll keep the assert independent. And maybe open a new PR for the helper function.

**Comment by sgugger** [comment]
I must admit I do not understand what the problem is, since the notebook linked executes without any issue.

**Comment by CakeCrusher** [comment]
Sorry about that @sgugger the notebook was organized in a weird way. Now [the notebook](https://colab.research.google.com/drive/16aLu6QrDSV_aUYRdpufl5E4iS08qkUGj?usp=sharing) will raise the error.

**Comment by sgugger** [comment]
I see. I've pointed out in #18119 where that error message should be updated.

**Pull request closed without merging**